src/provider/provider_tracking.c

```diff
@@ -119,6 +120,134 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
     return ret;
 }
 
+static umf_result_t trackingAllocationSplit(void *provider, void *ptr,
+                                            size_t totalSize, size_t leftSize) {
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *splitValue =
+        (tracker_value_t *)malloc(sizeof(tracker_value_t));
+    splitValue->pool = p->pool;
+    splitValue->size = leftSize;
+
+    tracker_value_t *value =
+        (tracker_value_t *)critnib_get((critnib *)p->hTracker, (uintptr_t)ptr);
+    if (!value) {
+        free(splitValue);
+        fprintf(stderr, "tracking split: no such value\n");
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+    if (value->size != totalSize) {
+        free(splitValue);
+        fprintf(stderr, "tracking split: %zu != %zu\n", value->size, totalSize);
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_result_t ret = umfMemoryProviderAllocationSplit(p->hUpstream, ptr,
+                                                        totalSize, leftSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        free(splitValue);
+        fprintf(stderr,
+                "tracking split: umfMemoryProviderAllocationSplit failed\n");
+        return ret;
+    }
+
+    void *rightPtr = (void *)(((uintptr_t)ptr) + leftSize);
+    size_t rightSize = totalSize - leftSize;
+
+    // We'll have duplicate entry for the range [rightPtr, rightValue->size] but this is fine,
+    // the value is the same anyway and we forbid splitting/removing that range concurrently
+    ret = umfMemoryTrackerAdd(p->hTracker, p->pool, rightPtr, rightSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        free(splitValue);
+        fprintf(stderr, "tracking split: umfMemoryTrackerAdd failed\n");
+        // TODO: what now? should we rollback the split? This is can only happen due to ENOMEM
+        // so it's unlikely but probably the best solution would be to try to preallocate everything
+        // (value and critnib nodes) before calling umfMemoryProviderAllocationSplit.
+        return ret;
+    }
+
+    int cret = critnib_insert((critnib *)p->hTracker, (uintptr_t)ptr,
+                              (void *)splitValue, 1 /* update */);
+    // this cannot fail since we know the element exists (nothing to allocate)
+    assert(cret == 0);
+    (void)cret;
+
+    // free the original value
+    free(value);
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static umf_result_t trackingAllocationMerge(void *provider, void *leftPtr,
+                                            void *rightPtr, size_t totalSize) {
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *mergedValue =
+        (tracker_value_t *)malloc(sizeof(tracker_value_t));
```

Comment:
if you allocate `mergedValue` just before `critnib_insert` then you do not need to `free(mergedValue)` in every error case below.

Reply:
Make sense, but in case of merge do we really need to allocate a new element? You can just update the left record with the new size and remove the right record. Did I miss something?

Reply:
See my other answer about updating the values. Without this alloc we would have to modify the value in-place which means we would have to use atomic instructions.